### PR TITLE
Remove metadata-proxy from GCE Clusters from 1.29+ onwards

### PR DESCRIPTION
**Pull request by upodroid**
/cc @justinsb 

More details for why: https://kubernetes.slack.com/archives/C8MKE2G5P/p1694044152804789

### CODE REVIEW

File: pkg/apis/kops/cluster.go [hunk 1]
```diff
@@ -244,6 +244,8 @@ type GCESpec struct {
 	NodeInstancePrefix *string `json:"nodeInstancePrefix,omitempty"`
 	// PDCSIDriver is the config for the PD CSI driver.
 	PDCSIDriver *PDCSIDriver `json:"pdCSIDriver,omitempty"`
+	// MetadataProxy is the config for the metadata-proxy
```
Comment (johngmyers) [inline pkg/apis/kops/cluster.go:247]:
```suggestion
	// MetadataProxy is the config for the metadata-proxy.
```

File: pkg/model/components/gcpmetadataproxy.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (johngmyers) [inline pkg/model/components/gcpmetadataproxy.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder/bootstrapchannelbuilder.go [hunk 3]
```diff
@@ -848,8 +848,8 @@ func (b *BootstrapChannelBuilder) buildAddons(c *fi.CloudupModelBuilderContext)
 	}
 
 	// The metadata-proxy daemonset conceals node metadata endpoints in GCE.
-	// It will land on nodes labeled cloud.google.com/metadata-proxy-ready=true
-	if b.Cluster.Spec.GetCloudProvider() == kops.CloudProviderGCE {
+	// It will land on nodes labeled cloud.google.com/metadata-proxy-ready=true when the feature is enabled
+	if b.Cluster.Spec.CloudProvider.GCE != nil && b.Cluster.Spec.CloudProvider.GCE.MetadataProxy != nil && fi.ValueOf(b.Cluster.Spec.CloudProvider.GCE.MetadataProxy.Enabled) {
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder/bootstrapchannelbuilder.go:852]: MetadataProxy is guaranteed to be non-`nil` at this point. But if it were `nil` then the proxy would be enabled.

Same for MetadataProxy.Enabled.

**Comment by upodroid** [comment]
Changing the PR to remove the feature entirely as discussed during the office hours. GCP deprecated this [product](https://cloud.google.com/kubernetes-engine/docs/how-to/protecting-cluster-metadata#concealment) a long time ago from GKE and the source code for it isn't around.

**Comment by johngmyers** [comment]
The Office Hours decision was to remove it as of Kubernetes 1.29. If the Kubernetes version is 1.28 or earlier then it should still be installed.

Do we need to add any code to get it removed upon upgrade over the 1.29 boundary?

**Comment by upodroid** [comment]
Fixed

**Comment by upodroid** [comment]
Verified the changes in the `pull-kops-e2e-k8s-gce-ci` job which is testing against 1.29

**Comment by upodroid** [comment]
These jobs should have focus_regex to Conformance

File: docs/releases/1.28-NOTES.md [hunk 4]
```diff
@@ -12,6 +12,8 @@ This is a document to gather the release notes prior to the release.
 
 ## GCP
 
+* We no longer deploy [metadata-proxy](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/metadata-proxy) on GCE clusters for K8s versions higher than 1.28. The proxy used to run as a DaemonSet and landed on nodes with the nodeLabel `cloud.google.com/metadata-proxy-ready: "true"`.
```
Comment (hakman) [inline docs/releases/1.28-NOTES.md:15]:
```suggestion
* [metadata-proxy](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/metadata-proxy) is no longer deployed on GCP clusters for Kubernetes 1.28+.
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by hakman** [comment]
> These jobs should have focus_regex to Conformance

We can debate this at the next kOps Office Hours. 😆

**Comment by upodroid** [comment]
I'm seeing infra bugs

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kops/15885/pull-kops-e2e-k8s-gce-ci/1700408126739582976/build-log.txt

my-hostname-private-cc309a20-27eb-4138-8dd2-b86c8864d452-j7pjv

image

**Comment by upodroid** [comment]
Fixed the docs

**Review by hakman — APPROVED** [review]

**Pull request merged**